File: client/web/model/shipyard.dart
part of kcaa_model;

class BuildSlot extends Observable {
  static final int STATE_EMPTY = 0;
  static final int STATE_BUILDING = 2;
  static final int STATE_COMPLETED = 3;
  static final Map<int, String> STATE_MAP = <int, String>{
    STATE_EMPTY: "空き",
    STATE_BUILDING: "建造中",
    STATE_COMPLETED: "建造完了",
  };

  @observable int id;
  @observable Ship ship;
  @observable int fuelSpent, ammoSpent, steelSpent, bauxiteSpent;
  @observable int material;
  @observable int state;
  @observable String stateString;
  @observable bool empty;
  @observable DateTime eta;
  @observable String etaDatetimeString;
  @observable bool completed;

  BuildSlot();

  void update(Map<String, dynamic> data, Map<int, Ship> shipDefinitionMap) {
    id = data["id"];
    ship = shipDefinitionMap[data["ship_def_id"]];
    fuelSpent = data["spent_resource"]["fuel"];
    ammoSpent = data["spent_resource"]["ammo"];
    steelSpent = data["spent_resource"]["steel"];
    bauxiteSpent = data["spent_resource"]["bauxite"];
    material = data["material"];
    state = data["state"];
    stateString = STATE_MAP[state];
    empty = data["empty"];

    // ETA.
    if (data["eta"] != 0) {
      eta = new DateTime.fromMillisecondsSinceEpoch(data["eta"], isUtc: true)
          .toLocal();
      etaDatetimeString = formatShortTime(eta);
    } else {
      eta = null;
      etaDatetimeString = null;
    }
    completed = state == STATE_COMPLETED ||
        (eta != null && eta.isBefore(new DateTime.now()));
  }
}

File: server/kcaa/kcsapi/shipyard.py
#!/usr/bin/env python

import jsonobject
import model
import resource


class BuildSlot(jsonobject.JSONSerializableObject):

    id = jsonobject.ReadonlyJSONProperty('id', value_type=int)
    """ID."""
    ship_def_id = jsonobject.JSONProperty('ship_def_id', value_type=int)
    """ID of ship definition being repaired in this slot."""
    spent_resource = jsonobject.JSONProperty(
        'spent_resource', value_type=resource.Resource)
    """Resource used to bulid this ship."""
    material = jsonobject.JSONProperty('material', value_type=int)
    """Building material used to build this ship."""
    state = jsonobject.JSONProperty('state', value_type=int)
    """State of the slot."""
    STATE_EMPTY = 0
    STATE_BULIDING = 2
    STATE_COMPLETED = 3
    eta = jsonobject.JSONProperty('eta', value_type=long)
    """Estimated Time of Arrival, in UNIX time with millisecond precision."""

    @jsonobject.jsonproperty
    def empty(self):
        return self.state == BuildSlot.STATE_EMPTY

    def completed(self, now):
        return (not self.empty and
                (self.state == BuildSlot.STATE_COMPLETED or now >= self.eta))


class BuildDock(model.KCAAObject):
    """Build dock."""

    slots = jsonobject.JSONProperty('slots', [], value_type=list,
                                    element_type=BuildSlot)
    """Build slots."""

    @property
    def empty_slots(self):
        return [slot for slot in self.slots if slot.empty]

    def update(self, api_name, request, response, objects, debug):
        super(BuildDock, self).update(api_name, request, response, objects,
                                      debug)
        if api_name == '/api_get_member/kdock':
            self.update_slots(response.api_data)
        elif api_name == '/api_req_kousyou/getship':
            self.update_slots(response.api_data.api_kdock)
        elif api_name == '/api_req_kousyou/createship_speedchange':
            slot = self.slots[int(request.api_kdock_id) - 1]
            slot.state = BuildSlot.STATE_COMPLETED

    def update_slots(self, kdock_data):
        self.slots = []
        for data in kdock_data:
            if data.api_state == -1:
                # Not opened.
                continue
            self.slots.append(BuildSlot(
                id=data.api_id,
                ship_def_id=data.api_created_ship_id,
                spent_resource=resource.Resource(
                    fuel=data.api_item1,
                    ammo=data.api_item2,
                    steel=data.api_item3,
                    bauxite=data.api_item4),
                material=data.api_item5,
                state=data.api_state,
                eta=long(data.api_complete_time)))
